apps/src/code-studio/initApp/loadApp.js

```diff
@@ -153,4 +337,84 @@ module.exports = function (callback) {
   } else {
     loadLastAttemptFromSessionStorage();
   }
+}
+
+window.dashboard = window.dashboard || {};
+window.dashboard.project = project;
+
+window.apps = {
+
+  // Set up projects, skipping blockly-specific steps. Designed for use
+  // by levels of type "external".
+  setupProjectsExternal: function () {
+    if (!window.dashboard) {
+      throw new Error('Assume existence of window.dashboard');
+    }
+  },
+
+  // Define blockly/droplet-specific callbacks for projects to access
+  // level source, HTML and headers.
+  sourceHandler: {
+    setInitialLevelHtml: function (levelHtml) {
+      appOptions.level.levelHtml = levelHtml;
+    },
+    getLevelHtml: function () {
+      return window.Applab && Applab.getHtml();
+    },
+    setInitialLevelSource: function (levelSource) {
+      appOptions.level.lastAttempt = levelSource;
+    },
+    // returns a Promise to the level source
+    getLevelSource: function (currentLevelSource) {
+      return new Promise((resolve, reject) => {
+        let source;
+        if (window.Blockly) {
+          // If we're readOnly, source hasn't changed at all
+          source = Blockly.mainBlockSpace.isReadOnly() ? currentLevelSource :
+                   Blockly.Xml.domToText(Blockly.Xml.blockSpaceToDom(Blockly.mainBlockSpace));
+          resolve(source);
+        } else if (appOptions.getCode) {
+          source = appOptions.getCode();
+          resolve(source);
+        } else if (appOptions.getCodeAsync) {
+          appOptions.getCodeAsync().then((source) => {
+            resolve(source);
+          });
+        }
+      });
+    },
+    setInitialAnimationList: function (animationList) {
+      appOptions.initialAnimationList = animationList;
+    },
+    getAnimationList: function (callback) {
+      if (appOptions.getAnimationList) {
+        appOptions.getAnimationList(callback);
+      } else {
+        callback({});
+      }
+    }
+  },
 };
+
+/**
+ * Sets up everything needed for an app (like applab, gamelab, eval, etc.) to run.
+ * This should only be called once per page load, with appoptions specified as a
+ * data attribute on the script tag.
+ */
+export default function loadApp(appMain) {
+  // Loads the dependencies for the current app based on values in `appOptions`.
+  // This function takes a callback which is called once dependencies are ready.
+  const script = document.querySelector(`script[data-appoptions]`);
+  const appOptions = JSON.parse(script.dataset.appoptions);
+  if (appOptions.embedded) {
+    // when we just "embed" an app (i.e. via embed_blocks.html.erb),
+    // we just want to call the appMain function immediately without
+    // loading a bunch of other stuff.
+    appMain(appOptions);
```

Comment:
it makes me cry that there is this very subtle special case controlled by an obscure flag that is far removed from the code that the flag controls.  Basically, passing down flags deep into the call stack to effectively do control flow in a way that doesn't make semantic sense is difficult to understand when you are reading the code.

Like, after I read the comment that I wrote above this line, I still don't understand why this case exists or is necessary, and if this code were well written, I wouldn't even need a comment to explain anything, it would just make sense.